etsy/phan
# Add PhanParamSignatureRealMismatch, which will completely ignore phpdoc types.

For issue #374

Use real signature types for the PhanSignatureReal* issues.
This ensures that we will catch more fatal errors - The php interpreter
completely ignores phpdoc types.

(E.g. something with only `/** @return void */` and no return type
is incompatible with a base class with a return type (` : void`), but PhanParamSignatureMismatch doesn't catch that)

TODO: test this on a codebase larger than Phan
EDIT: tested, didn't see any false positive errors (no issues saying there was a mismatch where there wasn't)
TODO: benchmark this
EDIT: self-analysis increased from 5.64s to 5.84s (single-threaded), which looks acceptable

- Change rules for that the checks of *real* signatures to be closer
  to what Zend/zend_inheritance.c checks in php-src
- Add tests, similar to those of Zend/zend_inheritance.c in php 7.1 
  (e.g. the tests of return type compatibility)
- In error messages, use the real types, not the phpdoc types.
- Test edge cases of inheriting traits (different PR?)
- Test edge cases of abstract methods (different PR?)
- Limit checks of `iterable` to `Traversable` and `array` to match
  PHP's checks for return types exactly.
- TODO: in another PR, separate the function signatures' minimum required args, and optional arg count? (Third-party modules are usually inaccurate here, but this is useful when extending)
